### PR TITLE
Refactored assignment of QC states

### PR DESCRIPTION
**Pull request by mgcam**

### CODE REVIEW

**Review by nerdstrike — CHANGES_REQUESTED** [review]
Big improvement, but I have concerns about code readability.

File: lang_qc/db/helper/well.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+class InconsistentInputError(Exception):
+    """
+    Custom exception for cases when individual values of attributes
+    are valid, but are inconsistent or mutually exlusive in regards
+    of the QC state that has to be assigned.
+    """
+
+
+class WellMetrics(BaseModel):
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/db/helper/well.py:72]: The fact that it is single use allows for caching of results of computation in cached properties

File: lang_qc/db/helper/well.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+        """
+        Given a description of the QC state, returns an object that
+        corresponds to a relevant row in the `qc_state_dict` table.
+        Raises a `ValidationError` if no corresponding dictionary record
+        is found.
+        """
+
+        valid = (
+            self.session.execute(
+                select(QcStateDict).where(QcStateDict.state == qc_state)
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/db/helper/well.py:168]: Good point

File: lang_qc/endpoints/pacbio_well.py [hunk 3]
```diff
@@ -134,122 +129,108 @@ def get_pacbio_well(
 
 @router.post(
     "/run/{run_name}/well/{well_label}/qc_claim",
-    summary="Well level QC operation - claim the well to start QC",
+    summary="Claim the well to start QC",
+    description="""
+    Enables the user to initiate manual QC of the well. The resulting QC state
+    is flagged as preliminary. Not every authenticated user is allowed to perform QC.
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/endpoints/pacbio_well.py:135]: Yes it is. The user should be listed in the `user` table.

File: lang_qc/models/qc_state.py [hunk 4]
```diff
@@ -87,30 +98,19 @@ def from_orm(cls, obj: QcStateDB):
         A class factory method. Given a database object representing the
         QC state, returns an instance of this class object.
         """
+
+        # TODO: dates should be converted to local time, including DST
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/models/qc_state.py:102]: Agree

**Comment by nerdstrike** [comment]
Naïvely the full stack seems to work with this set of changes.

File: lang_qc/db/helper/well.py [hunk 5]
```diff
@@ -110,94 +110,114 @@ def exists(self) -> bool:
...
+        keep_untouched = (cached_property,)
+
+    @cached_property
+    def seq_product(self) -> SeqProduct:
         """
-        Given a description of the QC type, returns an object that
-        corresponds to a relevant row in the `qc_type` table.
-        Raises a `ValidationError` if no corresponding record is found.
+        A cached instance of the lang_qc.db.qc_schema.SeqProduct object.
+        If the corresponding database record exists, this object corresponds
```
Comment (nerdstrike) [inline lang_qc/db/helper/well.py:196]: This is not an object, it's a property. Therefore more simply "Returns a pre-existing SeqProduct, or creates a new one."?

**Review by nerdstrike — CHANGES_REQUESTED** [review]
A few more docstring troubles.

**Review by nerdstrike — APPROVED** [review]

**Pull request merged**